### ISSUE TITLE
Assorted feature requests

### ISSUE THREAD

**Issue by justheuristic** [learning-at-home/hivemind#6]
This issue contains a list of features that are nice to have but not required immediately. Feel free to add stuff, edit stuff or convert stuff into working code :)
If you're interested, please leave a comment first.

### ~Runtime: ReplicaExpertBackend~
Update: this idea was generalized in https://openreview.net/forum?id=U1edbV4kNu_ , we will merge the extended version.

Right now every expert is served by at most one server. If we have 100500 GPUs but we don't want that many experts, we can implement asynchronous data parallelism within each expert.

Design idea: if an expert is already present in the network as an ExpertBackend, let others replicate it via some kind of ReplicaExpertBackend. Replica periodically requests expert's parameters and sends accumulated gradients with some staleness information. In turn, master ExpertBackend incorporates these gradients into training. This will improve performance if we load updates and send accumulated gradients infrequently enough. 

### ~Server: Dynamic expert loading~
Update: this idea was generalized in https://openreview.net/forum?id=U1edbV4kNu_ , we will merge the extended version.

Currently all hivemind servers start with a fixed number of experts and maintain these experts until the server is shut down. Ideally, we could dynamically load most promising experts from the network and accommodate them on the current server.

Can we pose it as some optimization procedure that is performed locally from every server? For instance, let's say that each gating function call "rewards" chosen experts proportionally to their weight in the mixture for that call's input - but only if the expert returned something.

In that case, we could design HivemindServer to optimize its total reward. For instance, a server can periodically replace low-reward experts with the ones that have highest expected reward per {second? per computation? depending on server status}. If replication is implemented, a server can choose between hosting a new expert or sharing a 1/N-th of a more popular expert.

Perhaps we can even re-think HivemindServer's behavior as optimization of the expected loss function?

### ~Server: Dashboard~
Update: @mryab and @borzunov significantly improved the way we do CLI logs. It seems that implementing a visual dashboard is not a priority now.
Debugging decentralized systems is easier if you can view the history of what happened. 

We could use some user-friendly way of logging server load, latency, the number of errors, etc.
Currently no such thing exists. Ideally, we should use something like tensorboard/graphana as a monitoring tool.

### ~More responsive communication between RemoteExpert and HivemindServer~
_Suggested by Dmitriy Afananasiev_
Update: will attempt in sub-projects, summer 2022

Right now HivemindServer accepts requests and goes dark until they are processed. Instead, we could instantly comunicate some status response e.g. "received request, i will answer in approx. T seconds". Using this, a RemoteExpert could choose to wait for response or cancel it right away (or choose a different replica).

### ~Richer DHT.get with subkeys~
Updatge: Implemented a while ago.
Right now, calling DHT.get for a value with subkeys has two potential outcomes:
* latest=True -> fetch all subkeys with latest expiration time, take your time
* latest=False -> gimme something - maybe just one subkey, maybe someone else's cached subkey

It would be convenient to also have an option to
* _required_subkeys_ - do not finish search until you find all of these subkeys
* min_subkeys - do not finish until you find a certain number of subkeys with sufficient expiration time
* while we're at it, DHTNode cache refresh should also re-request the existing subkeys

**Comment by justheuristic** [comment]
Replicas and dynamic expert loading were implemented in SWARM and will be measured within @mryab 's effort of merging SWARM.

Subkeys were implemented a while ago.

Responsive hivemind.Server api is currently tested in an internal sub-project: if it ends up useful, we'll merge it.
